hooks/open-telemetry/pkg/otel.go

```diff
@@ -2,165 +2,53 @@ package otel
-	h.ctx = ctx
-}
-
-// Before creates the flag evaluations open-telemetry span and sets the FlagKey and ProviderName attributes
-func (h *Hook) Before(hookContext of.HookContext, hookHints of.HookHints) (*of.EvaluationContext, error) {
-	key := fmt.Sprintf("%s.%s", hookContext.ClientMetadata().Name(), hookContext.FlagKey())
-	h.setup()
-	if _, ok := h.spans[key]; !ok {
-		h.spans[key] = &mutexWrapper{
-			mu: &sync.Mutex{},
-		}
-	}
-	h.spans[key].mu.Lock()
-	h.wg.Add(1)
-	ctx, span := h.tracerClient.tracer().Start(h.ctx, key)
-	ctx, cancel := context.WithCancel(ctx)
-	span.SetAttributes(
-		attribute.String(AttributeFlagKey, hookContext.FlagKey()),
-		attribute.String(AttributeProviderName, hookContext.ProviderMetadata().Name),
-	)
-	h.spans[key].ss = &storedSpan{
-		cancel: cancel,
-		span:   span,
+		ctx: ctx,
 	}
-	// this goroutine cleans up the span, if the associated context is closed then the stored
-	// span data is removed and the resource is unlocked. This context close can come from either the cancel() method
-	// or from the closing of the parent context outside the scope of the hook
-	go func() {
-		<-ctx.Done()
-		h.spans[key].ss = nil
-		h.spans[key].mu.Unlock()
-		h.wg.Done()
-	}()
-	return nil, nil
 }
 
-// After sets the EvaluatedVariant and EvaluatedValue on the evaluation specific span
-func (h *Hook) After(hookContext of.HookContext, flagEvaluationDetails of.InterfaceEvaluationDetails, hookHints of.HookHints) error {
-	key := fmt.Sprintf("%s.%s", hookContext.ClientMetadata().Name(), hookContext.FlagKey())
-	mw, ok := h.spans[key]
-	if !ok {
-		return errors.New("no span stored for provided hook context")
-	}
-	if flagEvaluationDetails.ResolutionDetail.Variant != "" {
-		mw.ss.span.SetAttributes(
-			attribute.String(AttributeEvaluatedVariant, flagEvaluationDetails.ResolutionDetail.Variant),
-		)
+// After sets the feature_flag event and associated attributes on the span stored in the context
+func (h *Hook) After(hookContext openfeature.HookContext, flagEvaluationDetails openfeature.InterfaceEvaluationDetails, hookHints openfeature.HookHints) error {
+	if h.ctx == nil {
+		// if no context is set trace.SpanFromContext will return a noop span
 		return nil
 	}
-	var value string
-	switch flagEvaluationDetails.FlagType {
-	case of.Boolean:
-		value = fmt.Sprintf("%t", flagEvaluationDetails.Value.(bool))
-	case of.String:
-		value = flagEvaluationDetails.Value.(string)
-	case of.Float:
-		value = fmt.Sprintf("%64f", flagEvaluationDetails.Value.(float64))
-	case of.Int:
-		value = fmt.Sprintf("%d", flagEvaluationDetails.Value.(int64))
-	case of.Object:
-		val, err := json.Marshal(flagEvaluationDetails.Value)
-		if err != nil {
-			return err
-		}
-		value = string(val)
-	default:
-		return fmt.Errorf("unknown data type received: %d", flagEvaluationDetails.FlagType)
-	}
-	mw.ss.span.SetAttributes(
-		attribute.String(AttributeEvaluatedValue, value),
+	span := trace.SpanFromContext(h.ctx)
+	span.SetAttributes(
```

Comment:
This doesn't seem correct. The attributes should be associated with the span event not the span.